server/services/omnichannel-voip/service.ts

```diff
@@ -52,24 +104,110 @@ export class OmnichannelVoipService extends ServiceClass implements IOmnichannel
+		return allocatedExtensions.map((user: any) => ({
+			_id: user._id,
+			agentName: user.username,
+			extension: user.extension,
+		}));
+	}
+
+	/* Voip calls */
+	async getNewRoom(
+		guest: ILivechatVisitor,
+		agent: { agentId: string; username: string },
+		rid: string,
+		options: FindOneOptions<IVoipRoom> = {},
+	): Promise<IRoomCreationResponse> {
+		this.logger.debug(`Attempting to find or create a room for visitor ${guest._id}`);
+		let room = await this.voipRoom.findOneById(rid, options);
+		let newRoom = false;
+		if (room && !room.open) {
+			this.logger.debug(`Last room for visitor ${guest._id} closed. Creating new one`);
+			room = null;
+		}
+		if (room == null) {
+			const name = guest.name || guest.username;
+			const roomId = await this.createVoipRoom(rid, name, agent, guest);
+			room = await this.voipRoom.findOneVoipRoomById(roomId);
+			newRoom = true;
+			this.logger.debug(`Room obtained for visitor ${guest._id} -> ${room?._id}`);
+		}
+		if (!room) {
+			this.logger.debug(`Visitor ${guest._id} trying to access another visitor's room`);
+			throw new Error('cannot-access-room');
+		}
 		return {
-			result: allocatedExtensions.map((user: any) => ({
-				_id: user._id,
-				agentName: user.username,
-				extension: user.extension,
-			})),
+			room,
+			newRoom,
+		};
+	}
+
+	async findRoom(token: string, rid: string): Promise<IVoipRoom | null> {
+		const projection = {
+			t: 1,
+			departmentId: 1,
+			servedBy: 1,
+			open: 1,
+			v: 1,
+			ts: 1,
+		};
+		if (!rid) {
+			return this.voipRoom.findOneByVisitorToken(token, { projection });
+		}
+		return this.voipRoom.findOneByIdAndVisitorToken(rid, token, { projection });
+	}
+
+	async closeRoom(visitor: ILivechatVisitor, room: IVoipRoom /* , comment: any, options = {}*/): Promise<boolean> {
+		this.logger.debug(`Attempting to close room ${room._id}`);
+		if (!room || room.t !== 'v' || !room.open) {
+			return false;
+		}
+
+		// const params = callbacks.run('livechat.beforeCloseRoom', { room, options });
+		// const { extraData } = params;
+
+		const now = new Date();
+		const { _id: rid } = room;
+		const closer = visitor ? ('visitor' as const) : ('user' as const);
+		const closeData: IRoomClosingInfo = {
+			closedAt: now,
+			// TODO: calculate actual call duration
+			callDuration: now.getTime() / 1000,
+			closer,
+			// ...extraData,
 		};
+		this.logger.debug(`Room ${room._id} was closed at ${closeData.closedAt} (duration ${closeData.callDuration})`);
+
+		/*
+		// We should be able to handle nearend and farend call end.
+		if (user) {
+			this.logger.debug(`Closing by user ${ user._id }`);
+			closeData.closer = 'user';
+			closeData.closedBy = {
+				_id: user._id,
+				username: user.username,
+			};
+		} else
+		*/
+		if (visitor) {
+			this.logger.debug(`Closing by visitor ${visitor._id}`);
+			closeData.closedBy = {
+				_id: visitor._id,
+				username: visitor.username,
+			};
+		}
+
+		this.voipRoom.closeByRoomId(rid, closeData);
+		return true;
```

Comment:
Remove the commented code if it's not going to be used.

Reply:
Fixed.